### PR TITLE
Enhance Japanese language detection

### LINKED ISSUE #3565: [v1.1.0-rc.0] Japanese documents cannot be searched properly with Kanji-only documents
**Describe the bug**

I was testing the search by adding Japanese documents, but when I added a Kanji-only document, past Japanese documents were no longer caught in the search.

**To Reproduce**

summary

1. add documents `[{id: 1, name: "東京スカパラダイスオーケストラ"}]` -> search with `東京` : expected
2. add documents `[{id: 2, name: "東京バナナ"}, {id:3, name: "東京 ポテチ"}]` -> search with `東京` : expected
3. add documents `[{id: 4, name: "東京特許許可局"}]` -> search with `東京` : **not expected**

launch

```bash
docker run --rm -p 7700:7700 -e MEILI_MASTER_KEY=foobar getmeili/meilisearch:v1.1.0-rc.0
```

processing

```bash
# create an index
curl -H "Authorization: Bearer foobar" -H "Content-Type: application/json" -X POST localhost:7700/indexes -d '{"uid":"hogehoge","primaryKey":"id"}'

# add a Japanese document
curl -H "Authorization: Bearer foobar" -H "Content-Type: application/json" -X POST localhost:7700/indexes/hogehoge/documents -d '{"id":1,"name":"東京スカパラダイスオーケストラ"}'

# search and hit 1 document ( correct )
curl -H "Authorization: Bearer foobar" -H "Content-Type: application/json" -X POST localhost:7700/indexes/hogehoge/search -d '{"q":"東京"}'

{"hits":[{"id":1,"name":"東京スカパラダイスオーケストラ"}],"query":"東京","processingTimeMs":0,"limit":20,"offset":0,"estimatedTotalHits":1}

# add more documents
curl -H "Authorization: Bearer foobar" -H "Content-Type: application/json" -X POST localhost:7700/indexes/hogehoge/documents -d '{"id":2,"name":"東京バナナ"}'
curl -H "Authorization: Bearer foobar" -H "Content-Type: application/json" -X POST localhost:7700/indexes/hogehoge/documents -d '{"id":3,"name":"東京 ポテチ"}'

# search and hit 3 documents ( correct ) 
curl -H "Authorization: Bearer foobar" -H "Content-Type: application/json" -X POST localhost:7700/indexes/hogehoge/search -d '{"q":"東京"}'

{"hits":[{"id":1,"name":"東京スカパラダイスオーケストラ"},{"id":2,"name":"東京バナナ"},{"id":3,"name":"東京 ポテチ"}],"query":"東京","processingTimeMs":0,"limit":20,"offset":0,"estimatedTotalHits":3}

# add a Kanji only document
curl -H "Authorization: Bearer foobar" -H "Content-Type: application/json" -X POST localhost:7700/indexes/hogehoge/documents -d '{"id":4,"name":"東京特許許可局"}'

# search and hit only last document ( not expected )
curl -H "Authorization: Bearer foobar" -H "Content-Type: application/json" -X POST localhost:7700/indexes/hogehoge/search -d '{"q":"東京"}'

{"hits":[{"id":4,"name":"東京特許許可局"}],"query":"東京","processingTimeMs":0,"limit":20,"offset":0,"estimatedTotalHits":1}

# add a document again
curl -H "Authorization: Bearer foobar" -H "Content-Type: application/json" -X POST localhost:7700/indexes/hogehoge/documents -d '{"id":5,"name":"東京スカイツリー"}'

# search but hit only 1 document ( not expected )
curl -H "Authorization: Bearer foobar" -H "Content-Type: application/json" -X POST localhost:7700/indexes/hogehoge/search -d '{"q":"東京"}'

{"hits":[{"id":4,"name":"東京特許許可局"}],"query":"東京","processingTimeMs":0,"limit":20,"offset":0,"estimatedTotalHits":1}

```

**Expected behavior**

I expected all documents could searched by  `東京`

**Meilisearch version: v1.1.0-rc.0**

**Additional context**

Ref #3347

### PR DESCRIPTION
**Pull request by ManyTheFish**
# Pull Request

This PR is a prototype and can be tested by downloading [the dedicated docker image](https://hub.docker.com/layers/getmeili/meilisearch/prototype-better-language-detection-0/images/sha256-a12847de00e21a71ab797879fd09777dadcb0881f65b5f810e7d1ed434d116ef?context=explore):

```bash
$ docker pull getmeili/meilisearch:prototype-better-language-detection-0
```

## Context
Some Languages are harder to detect than others, this miss-detection leads to bad tokenization making some words or even documents completely unsearchable. Japanese is the main Language affected and can be detected as Chinese which has a completely different way of tokenization.

A [first iteration has been implemented for v1.1.0](https://github.com/meilisearch/meilisearch/pull/3347) but is an insufficient enhancement to make Japanese work. This first implementation was detecting the Language during the indexing to avoid bad detections during the search.
Unfortunately, some documents (shorter ones) can be wrongly detected as Chinese running bad tokenization for these documents and making possible the detection of Chinese during the search because it has been detected during the indexing.

For instance, a Japanese document `{"id": 1, "name": "東京スカパラダイスオーケストラ"}` is detected as Japanese during indexing, during the search the query `東京` will be detected as Japanese because only Japanese documents have been detected during indexing despite the fact that v1.0.2 would detect it as Chinese.
However if in the dataset there is at least one document containing a field with only Kanjis like:
_A document with only 1 field containing only Kanjis:_
```json
{
 "id":4,
 "name": "東京特許許可局"
}
```
_A document with 1 field containing only Kanjis and 1 field containing several Japanese characters:_
```json
{
 "id":105,
 "name": "東京特許許可局",
 "desc": "日経平均株価は26日 に約8カ月ぶりに2万4000円の心理的な節目を上回った。株高を支える材料のひとつは、自民党総裁選で3選を決めた安倍晋三首相の経済政策への期待だ。恩恵が見込まれるとされる人材サービスや建設株の一角が買われている。ただ思惑が先行して資金が集まっている面 は否めない。実際に政策効果を取り込む企業はどこか、なお未知数だ。"
}
```

Then, in both cases, the field `name` will be detected as Chinese during indexing allowing the search to detect Chinese in queries. Therefore,  the query `東京` will be detected as Chinese and only the two last documents will be retrieved by Meilisearch.

## Technical Approach

The current PR partially fixes these issues by:
1) Adding a check over potential miss-detections and rerunning the extraction of the document forcing the tokenization over the main Languages detected in it.
 >  1) run a first extraction allowing the tokenizer to detect any Language in any Script
 >  2) generate a distribution of tokens by Script and Languages (`script_language`)
 >  3) if for a Script we have a token distribution of one of the Language that is under the threshold, then we rerun the extraction forbidding the tokenizer to detect the marginal Languages
 >  4) the tokenizer will fall back on the other available Languages to tokenize the text. For example, if the Chinese were marginally detected compared to the Japanese on the CJ script, then the second extraction will force Japanese tokenization for CJ text in the document. however, the text on another script like Latin will not be impacted by this restriction.

2) Adding a filtering threshold during the search over Languages that have been marginally detected in documents

## Limits
This PR introduces 2 arbitrary thresholds:
1) during the indexing, a Language is considered miss-detected if the number of detected tokens of this Language is under 10% of the tokens detected in the same Script (Japanese and Chinese are 2 different Languages sharing the "same" script "CJK").
2) during the search, a Language is considered marginal if less than 5% of documents are detected as this Language.

This PR only partially fixes these issues:
- ✅ the query `東京` now find Japanese documents if less than 5% of documents are detected as Chinese.
- ✅ the document with the id `105` containing the Japanese field `desc` but the miss-detected field `name` is now completely detected and tokenized as Japanese and is found with the query `東京`.
- ❌ the document with the id `4` no longer breaks the search Language detection but continues to be detected as a Chinese document and can't be found during the search.

## Related issue
Fixes #3565

## Possible future enhancements
- Change or contribute to the Library used to detect the Language
  - the related issue on Whatlang: https://github.com/greyblake/whatlang-rs/issues/122

### CODE REVIEW

**Comment by ManyTheFish** [comment]
Hey @dureuill,
I'd like your POV on this PR, thank you if you have the time to look at it

**Comment by dureuill** [comment]
Hello @ManyTheFish, thanks for thinking of me to review your PR!

Before I dive into the code, a question: did we consider the option of having some documents left as "language-ambiguous" and tokenized as many time as necessary in each of the candidate languages? Similarly, for search queries that could match several languages, tokenize them in all of these languages and return results for all the resulting queries?

Such a change could cause structural change and so be a large undertaking. Still, if it has been discussed somewhere, I'd love to be pointed to that discussion.

Meanwhile, the change proposed here appears like a nice heuristic to prevent the misdetection of languages using frequency.

I'm not 100% sure of what the scope of this sentence is:

> rerunning the extraction of the document forcing the tokenization over the main Languages detected

can you maybe rephrase a bit?

**Comment by ManyTheFish** [comment]
Hello @dureuill, I added a section in **Technical Approach** detailing the indexing process and giving an example at the end. Let me know if you need more! 😄

File: milli/src/update/index_documents/extract/extract_docid_word_positions.rs [hunk 1]
```diff
@@ -45,61 +47,133 @@ pub fn extract_docid_word_positions<R: io::Read + io::Seek>(
...
+
+                script_language_word_count.clear();
+
+                // rerun the extraction.
+                extract_tokens_from_document(
+                    &obkv,
+                    searchable_fields,
+                    &tokenizer,
+                    max_positions_per_attributes,
+                    &mut key_buffer,
```
Comment (dureuill) [inline milli/src/update/index_documents/extract/extract_docid_word_positions.rs:108]: When we rerun the extraction, we do not clear the `key_buffer` again. Could that be an issue?

File: milli/src/update/index_documents/extract/extract_docid_word_positions.rs [hunk 2]
```diff
@@ -183,3 +257,36 @@ fn process_tokens<'a>(
         })
         .filter(|(_, t)| t.is_word())
 }
+
+fn potential_language_detection_error(languages_frequency: &Vec<(Language, usize)>) -> bool {
```
Comment (dureuill) [inline milli/src/update/index_documents/extract/extract_docid_word_positions.rs:261]:
```suggestion
fn potential_language_detection_error(languages_frequency: &[(Language, usize)]) -> bool {
```

I'm a bit surprised this is not flagged by clippy

File: milli/src/update/index_documents/extract/extract_docid_word_positions.rs [hunk 3]
```diff
@@ -183,3 +257,36 @@ fn process_tokens<'a>(
...
+            None
+        } else {
+            Some((*script, languages))
+        }
+    } else {
+        None
+    }
+}
+
+fn compute_laguage_frequency_threshold(languages_frequency: &[(Language, usize)]) -> usize {
```
Comment (dureuill) [inline milli/src/update/index_documents/extract/extract_docid_word_positions.rs:289]:
```suggestion
fn compute_language_frequency_threshold(languages_frequency: &[(Language, usize)]) -> usize {
```

File: milli/src/update/index_documents/extract/extract_docid_word_positions.rs [hunk 4]
```diff
@@ -183,3 +257,36 @@ fn process_tokens<'a>(
...
+    let total: usize = languages_frequency.iter().map(|(_, c)| c).sum();
+    total / 10 // 10% is a completely arbitrar value.
```
Comment (dureuill) [inline milli/src/update/index_documents/extract/extract_docid_word_positions.rs:291]:
```suggestion
    total / 10 // 10% is a completely arbitrary value.
```

**Comment by dureuill** [comment]
Thank you Many it is indeed clearer with the added explanation!

File: meilisearch/src/routes/indexes/mod.rs [hunk 5]
```diff
@@ -100,6 +100,8 @@ pub async fn list_indexes(
             Ok(Some(IndexView::new(uid.to_string(), index)?))
         })?;
     // Won't cause to open all indexes because IndexView doesn't keep the `Index` opened.
+    // error when trying to fix it: the trait `ExactSizeIterator` is not implemented for `Flatten<IntoIter<Option<IndexView>>>`
+    #[allow(clippy::needless_collect)]
```
[the remark this answers is not in the document]
Comment (ManyTheFish) [reply meilisearch/src/routes/indexes/mod.rs:104]: I can remove it but I changed it because of the CI if I remind. Let's try it

File: milli/src/update/index_documents/extract/extract_docid_word_positions.rs (hunk 1 above)
Comment (ManyTheFish) [reply milli/src/update/index_documents/extract/extract_docid_word_positions.rs:108]: the `key_buffer` is the concatenation of the internal document id, which didn't change, and the field id, which is overwritten in the extractor.

File: milli/src/update/index_documents/extract/extract_docid_word_positions.rs (hunk 2 above)
Comment (ManyTheFish) [reply milli/src/update/index_documents/extract/extract_docid_word_positions.rs:261]: Because the function is used in an iterator:
```rust
if script_language_word_count.values().any(potential_language_detection_error) {
```

This would force me to change this implementation instead of directly passing the function. You may have a tip about this.

Comment (dureuill) [reply milli/src/update/index_documents/extract/extract_docid_word_positions.rs:257]: [Here's a playground with various options](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cc5e25c75985c768dc63ec9d68b811e2)

Comment (ManyTheFish) [reply milli/src/update/index_documents/extract/extract_docid_word_positions.rs:257]: I think that the solution with `map(Vec::as_slice)` is the more suitable, the others are way less readable for an idealistic way of coding 😬

Comment (dureuill) [reply milli/src/update/index_documents/extract/extract_docid_word_positions.rs:257]: That's a personal style question, I personally don't have anything against lambdas 😅

File: meilisearch/src/routes/indexes/mod.rs [hunk 6]
```diff
@@ -100,6 +100,7 @@ pub async fn list_indexes(
             Ok(Some(IndexView::new(uid.to_string(), index)?))
         })?;
     // Won't cause to open all indexes because IndexView doesn't keep the `Index` opened.
+    // error when trying to fix it: the trait `ExactSizeIterator` is not implemented for `Flatten<IntoIter<Option<IndexView>>>`
```
Comment (dureuill) [inline meilisearch/src/routes/indexes/mod.rs:103]: I think we can remove this line if clippy is not flagging it, otherwise it might be confusing

```suggestion
```

**Review by dureuill — APPROVED** [review]
Approving. There remain unresolved inline comments above that you can take or leave :-).

**Review by dureuill — APPROVED** [review]

File: milli/src/update/index_documents/extract/extract_docid_word_positions.rs (hunk 2 above)
Comment (ManyTheFish) [reply milli/src/update/index_documents/extract/extract_docid_word_positions.rs:257]: Well, I like lambdas, but sometimes, being able to read your code as a sentence is enjoyable, like:
`if in script_language_word_count values there is any potential_language_detection_error then do ...` 😋 😋

**Review by dureuill — APPROVED** [review]
perfect thank you :heart:

**Comment by dureuill** [comment]
bors merge
